### PR TITLE
Handle unwrapping of casts with partial injective coercion

### PR DESCRIPTION
**Pull request by rohangarg**

### CODE REVIEW

**Comment by rohangarg** [comment]
If a double literal in a cast comparison is > -2^53 and < 2^53, it will be injective w.r.t a bigint type. All bigint within this range are injective with double type in the specified range. The bigint values outside this range can imagined as infinite values which can still be compared to the constant.
If the double literal is outside this range, then the injective property is not valid and we can't do coercions.
Following rules can be used while doing the coercion : 
```
sourceType = BIGINT, targetType = DOUBLE (and T is < 2^53 and !NaN and !infinite)
if cast(cast(T, sourceType), targetType) = T:
	for all comparison operators c, we can rewrite the expression as:
		S c cast(T, sourceType)

if cast(cast(T, sourceType), targetType) > T (round up), we can rewrite the expression as:
   for > or >= operator:
       S >= cast(T, sourceType)
   for < or <= operator:
       S < cast(T, sourceType)
   for = operator: 
       false if S is non null
   for != operator:
   	   true if S is non null
   for is_distinct_from operator:
   	   true

if cast(cast(T, sourceType), targetType) < T (round down):
   for >, >= operator, we can rewrite the expression as:
       S > cast(T, sourceType)
   for <, <= operator, we can rewrite the expression as:
       S <= cast(T, sourceType)
   for = operator: 
       false if S is non null
   for != operator:
   	   true if S is non null
   for is_distinct_from operator:
   	   true
```
Following argument can also hold true for bigint->real type and integer->real type (the limit changes from 2^53 to 2^23 due to smaller mantissa)

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java [hunk 1]
```diff
@@ -309,7 +332,7 @@ private Expression unwrapCast(ComparisonExpression expression)
                         return TRUE_LITERAL;
                     case LESS_THAN:
                     case LESS_THAN_OR_EQUAL:
-                        if (sourceRange.isPresent() && compare(sourceType, sourceRange.get().getMin(), literalInSourceType) == 0) {
+                        if (sourceRange.isPresent() && compare(sourceType, sourceRange.get().getMin(), literalInSourceType) == 0 && implicitInjectiveCoercionAbility == FULL) {
```
Comment (kasiafi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java:335]: The additional condition is redundant. For PARTIAL, the allowed range for literal here is well within source type range. Instead, please add a comment.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java [hunk 2]
```diff
@@ -339,7 +362,7 @@ private Expression unwrapCast(ComparisonExpression expression)
                         return new ComparisonExpression(LESS_THAN, cast.getExpression(), literalEncoder.toExpression(literalInSourceType, sourceType));
                     case GREATER_THAN:
                     case GREATER_THAN_OR_EQUAL:
-                        if (sourceRange.isPresent() && compare(sourceType, sourceRange.get().getMax(), literalInSourceType) == 0) {
+                        if (sourceRange.isPresent() && compare(sourceType, sourceRange.get().getMax(), literalInSourceType) == 0 && implicitInjectiveCoercionAbility == FULL) {
```
Comment (kasiafi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java:365]: The additional condition is redundant. For PARTIAL, the allowed range for literal here is well within source type range. Instead, please add a comment.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java [hunk 3]
```diff
@@ -178,14 +181,34 @@ private Expression unwrapCast(ComparisonExpression expression)
...
+                        }
+                    }
+                    if (targetType.equals(REAL) && (sourceType.equals(BIGINT) || sourceType.equals(INTEGER))) {
+                        Float rightReal = Float.intBitsToFloat(((Long) right).intValue());
+                        if (rightReal.isInfinite() || rightReal.isNaN() || rightReal >= 1L << 23 || rightReal <= -1L << 23) {
+                            return expression;
+                        }
+                    }
+                }
+                catch (ClassCastException e) {
```
Comment (kasiafi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java:203]: Please add a type check to the if-conditions instead of try-catch.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java [hunk 4]
```diff
@@ -178,14 +181,34 @@ private Expression unwrapCast(ComparisonExpression expression)
...
             Type targetType = typeAnalyzer.getType(session, types, expression.getRight());
 
-            if (!hasInjectiveImplicitCoercion(sourceType, targetType)) {
+            ImplicitInjectiveCoercionAbility implicitInjectiveCoercionAbility = getImplicitInjectiveCoercionAbility(sourceType, targetType);
+            if (implicitInjectiveCoercionAbility == NONE) {
                 return expression;
             }
+            if (implicitInjectiveCoercionAbility == PARTIAL) {
+                try {
+                    if (targetType.equals(DOUBLE) && sourceType.equals(BIGINT)) {
```
Comment (kasiafi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java:190]: Flipping conjuncts would improve readability.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java [hunk 5]
```diff
@@ -178,14 +181,34 @@ private Expression unwrapCast(ComparisonExpression expression)
...
+                        Double rightDouble = (Double) right;
+                        if (rightDouble.isInfinite() || rightDouble.isNaN() || rightDouble >= 1L << 53 || rightDouble <= -1L << 53) {
+                            return expression;
```
Comment (kasiafi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java:193]: No need to give up here. Most of the cases could be handled:
- a comparison with NaN
- a comparison with any Double outside of Bigint range (including infinity)

Comment (rohangarg) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java:193]: Ok, yes. I think they can be handled too : 
1. for NaN, should we just remove the check? currently I think there is no optimization for NaN - we could add some check to detect NaN and return `false` predicate. 
2. if the double value is outside the bigint range, then probably we can do a `true/false` predicate depending on the operator. are you referring to that?

Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java:193]: > if the double value is outside the bigint range, then probably we can do a true/false predicate depending on the operator. are you referring to that?

yes, but mind the null. There's helper methods for that `falseIfNotNull()`, `trueIfNotNull()`.

> we could add some check to detect NaN and return false predicate.

That depends on the comparison operator, too. For NOT_EQUAL and for IS_DISTINCT_FROM  it will be true. And mind the null.

Comment (rohangarg) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java:193]: > yes, but mind the null. There's helper methods for that falseIfNotNull(), trueIfNotNull().

Ok, yes. Ideally, we can also let the optimizations done for boundary (or out of bound) values done in the rule. I think they can handle the case where double value >= 2^63 and <= -2^63. I will test that. 

>That depends on the comparison operator, too. For NOT_EQUAL and for IS_DISTINCT_FROM it will be true. And mind the null.

Yes, that's true.

Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java:193]: > Ok, yes. Ideally, we can also let the optimizations done for boundary (or out of bound) values done in the rule. I think they can handle the case where double value >= 2^63 and <= -2^63. I will test that.

Exactly. That I meant by "a comparison with any Double outside of Bigint range (including infinity)".

After this is done, the only uncovered areas will be those for the Double value ranging from (+/-) 2^35 until 2^63. Please add tests for all boundary values.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java [hunk 6]
```diff
@@ -309,6 +339,7 @@ private Expression unwrapCast(ComparisonExpression expression)
                         return TRUE_LITERAL;
                     case LESS_THAN:
                     case LESS_THAN_OR_EQUAL:
+                        // this condition will never be true for PARTIAL injective types
```
Comment (kasiafi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java:342]: This comment looks like a general rule for partial injective types, but it's not.
Now, that the out-of-bound cases are supported, it is not obvious that the condition is false. Actually, it is false because such cases would be handled / excluded until this point.
(see the comment on code structure)

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java [hunk 7]
```diff
@@ -339,6 +370,7 @@ private Expression unwrapCast(ComparisonExpression expression)
                         return new ComparisonExpression(LESS_THAN, cast.getExpression(), literalEncoder.toExpression(literalInSourceType, sourceType));
                     case GREATER_THAN:
                     case GREATER_THAN_OR_EQUAL:
+                        // this condition will never be true for PARTIAL injective types
```
Comment (kasiafi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java:373]: same as above

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java [hunk 8]
```diff
@@ -178,14 +181,41 @@ private Expression unwrapCast(ComparisonExpression expression)
             Type sourceType = typeAnalyzer.getType(session, types, cast.getExpression());
             Type targetType = typeAnalyzer.getType(session, types, expression.getRight());
 
-            if (!hasInjectiveImplicitCoercion(sourceType, targetType)) {
+            ImplicitInjectiveCoercionAbility implicitInjectiveCoercionAbility = getImplicitInjectiveCoercionAbility(sourceType, targetType);
+            boolean boundaryChecksAllowed = implicitInjectiveCoercionAbility == FULL;
```
Comment (kasiafi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java:185]: Instead of adding a boolean, I would restructure the code so that there are different methods for different unwrapping routines. It would be easier to follow the flow:
`FULL` --> `out-of-bound unwrapping` --> `in-bound unwrapping` --> give up
`PARTIAL` --> `analyse rightValue` --> either `out-of-bound unwrapping` or `in-bound unwrapping` or NaN or give up

**Comment by rohangarg** [comment]
@kasiafi : Apologies for the delay in the review fixes. I was occupied with something else recently. 
I have added some changes regarding restructuring, keeping them in separate commit for easier iteration for now.
The `NaN` case is still on my TODO, will do it very soon too.

**Comment by rohangarg** [comment]
@martint : I checked and the approach you mentioned seems correct to me. IIRC, I did the earlier changes to make the boundary checks run only for cases where the literal was infinite. But I missed that the boundary checking code can also be run for cases where the literal is (-2^53, 2^53) - boundary checks will not be incorrect due to rounded casting since the values are well within the boundaries.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java [hunk 9]
```diff
@@ -351,14 +352,25 @@ private Expression unwrapCast(ComparisonExpression expression)
...
                 // not every INTEGER fits in a REAL (32-bit vs 23-bit mantissa)
-                return false;
+                if (target.equals(DOUBLE)) {
+                    double doubleValue = (double) value;
+                    return doubleValue > Long.MAX_VALUE ||
+                            doubleValue < Long.MIN_VALUE ||
+                            (doubleValue > -1L << 53 && doubleValue < 1L << 53); // in (-2^53, 2^53), bigint follows an injective implicit coercion w.r.t double
+                }
+                else {
+                    float realValue = intBitsToFloat(((Long) value).intValue());
```
Comment (kasiafi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java:369]: Cast `value` to primitive long and use `toIntExact()` instead of `intValue()`.

Comment (rohangarg) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java]: done

File: presto-main/src/test/java/io/prestosql/sql/planner/TestUnwrapCastInComparison.java [hunk 10]
```diff
@@ -263,6 +299,30 @@ public void testLessThan()
...
+        assertPlan(
+                "SELECT * FROM (VALUES BIGINT '0') t(a) WHERE a < DOUBLE '1'",
+                anyTree(
+                        filter("A < BIGINT '1'",
+                                values("A"))));
+
+        assertPlan(
+                "SELECT * FROM (VALUES BIGINT '0') t(a) WHERE a < DOUBLE '1.1'",
+                anyTree(
+                        filter("A <= SMALLINT '1'",
```
[the remark this answers is not in the document]
Comment (rohangarg) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestUnwrapCastInComparison.java:312]: Yes, it should be `BIGINT` - but the test still passes. I checked and have a couple of findings : 
1. both the literals `BIGINT 1` (after the unwrap) and `SMALLINT 1` are converted to GenericLiterals when being converted to expression. bigints are converted to generic literals in when the value is in integer range, and the integers are rep as `LongLiterals`. I'm not sure about the reason behind not keeping integers as generic and bigints as longLiterals. could be due to some optimization or the timeline of the datatypes. 
2. when comparing two generic literals, `ExpressionVerifier` doesn't check their types. It just checks whether the values are same or not. Hence, the tests pass. I'm not sure whether this is intentional.

[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestUnwrapCastInComparison.java:312]: https://github.com/prestosql/presto/issues/3981

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java [hunk 11]
```diff
@@ -382,14 +382,27 @@ private Expression unwrapCast(ComparisonExpression expression)
...
                     (source.equals(BIGINT) && target.equals(REAL)) ||
                     (source.equals(INTEGER) && target.equals(REAL))) {
                 // Not every BIGINT fits in DOUBLE/REAL due to 64 bit vs 53-bit/23-bit mantissa. Similarly,
                 // not every INTEGER fits in a REAL (32-bit vs 23-bit mantissa)
-                return false;
+                if (target.equals(DOUBLE)) {
+                    double doubleValue = (double) value;
+                    return doubleValue > Long.MAX_VALUE ||
+                            doubleValue < Long.MIN_VALUE ||
+                            Double.isNaN(doubleValue) ||
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java:396]: Can you explain why we return true in these cases?

Comment (rohangarg) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java:396]: For cases where the double literal is greater (or less) than any possible bigint value when casted to a double, we can treat it as a beyond/below type boundary check for bigint type - example if the expression is like `bigint_col < cast(2^100 as double)`, then we can unwrap that case.
The NaN case is sent true because it gets unwrapped later as a part of NaN unwrap opt. done recently.

**Comment by martint** [comment]
Merged, thanks!
